helix-view/src/handlers/word_index.rs

```diff
@@ -0,0 +1,484 @@
+                // forget them and delete the old text.
+                if let Some(change) = self.changes.remove(&doc) {
+                    self.coordinator
+                        .send(Event::Delete(doc, change.old_text))
+                        .unwrap();
+                } else {
+                    self.coordinator.send(Event::Delete(doc, text)).unwrap();
+                }
+                timeout
+            }
+        }
+    }
+
+    fn finish_debounce(&mut self) {
+        for (doc, change) in self.changes.drain() {
+            self.coordinator.send(Event::Update(doc, change)).unwrap();
+        }
+    }
+}
+
+/// Minimum number of grapheme clusters required to include a word in the index
+const MIN_WORD_GRAPHEMES: usize = 3;
+/// Maximum word length allowed (in chars)
+const MAX_WORD_LEN: usize = 50;
+
+type Word = helix_stdx::str::TinyBoxedStr;
+
+#[derive(Debug, Default)]
+struct WordIndexInner {
+    /// Reference counted storage for words.
+    ///
+    /// Words are very likely to be reused many times. Instead of storing duplicates we keep a
+    /// reference count of times a word is used. When the reference count drops to zero the word
+    /// is removed from the index.
+    words: HashMap<Word, u32>,
+}
+
+impl WordIndexInner {
+    fn words(&self) -> impl Iterator<Item = &Word> {
+        self.words.keys()
+    }
+
+    fn insert(&mut self, word: RopeSlice) {
+        assert!(word.len_chars() <= MAX_WORD_LEN);
+        // The word must be shorter than `TinyBoxedStr::MAX` because it is fewer than 50
+        // characters and characters take at most four bytes.
+        assert!(word.len_bytes() < Word::MAX_LEN);
+
+        let word: Cow<str> = word.into();
+        if let Some(rc) = self.words.get_mut(word.as_ref()) {
+            *rc = rc.saturating_add(1);
+        } else {
+            self.words.insert(word.try_into().unwrap(), 1);
+        }
+    }
+
+    fn remove(&mut self, word: RopeSlice) {
+        let word: Cow<str> = word.into();
+        match self.words.get_mut(word.as_ref()) {
+            Some(1) => {
+                self.words.remove(word.as_ref());
+            }
+            Some(n) => *n -= 1,
+            None => (),
+        }
+    }
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct WordIndex {
+    inner: Arc<RwLock<WordIndexInner>>,
+}
+
+impl WordIndex {
+    pub fn matches(&self, pattern: &str) -> Vec<String> {
+        let inner = self.inner.read();
+        let mut matches = fuzzy_match(pattern, inner.words(), false);
+        matches.sort_unstable_by_key(|(_, score)| *score);
+        matches
+            .into_iter()
+            .map(|(word, _)| word.to_string())
+            .collect()
+    }
+
+    fn add_document(&self, text: &Rope) {
+        let words: Vec<_> = words(text.slice(..)).collect();
+        let mut inner = self.inner.write();
+        for word in words {
+            inner.insert(word);
+        }
```

Comment:
do we need the collect if `words()` already return an iterator ?

Reply:
This is so that we hold the write lock for only a very short time. I'm not sure it's necessary though as parsing the words out should be quite fast